tests/test_maintenance.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import io
+import fnmatch
+import unittest
+import unittest.mock
+import subprocess
+from tools.maintenance import maintenance
+
+DESCRIPTION = """
+This is a unit test script for the tools/maintenance/maintenance.py.
+
+Usage (run from top level directory of hpc toolkit):
+python -m unittest 
+"""
+
+# hpc-toolkit-test0 will have nodes that have maintenance
+# hpc-toolkit-test1 will not have nodes that have maintenance
+valid_prj_list = ["hpc-toolkit-test0", "hpc-toolkit-test1"]
+sched_maint_vms = [f"vm_test{x}" for x in range(10)]
+T1 = "2024-10-26T13:25:51-0700"
+T2 = "2024-10-26T13:26:51-0700"
+up_maint_vms = [[sched_maint_vms[0], T1, T2, "TRUE", "SCHEDULED"],
+                [sched_maint_vms[1], T1, T2, "FALSE", "SCHEDULED"],
+                [sched_maint_vms[2], T1, T2, "TRUE", "UNSCHEDULED"],
+                [sched_maint_vms[3], T1, T2, "FALSE", "UNSCHEDULED"]]
+version_res = """
+{
+    "alpha": "2023.11.10"
+}
+"""
+
+def subprocess_replace(cmd, shell, capture_output, text, 
+                       check) -> subprocess.CompletedProcess:
+    res = subprocess.CompletedProcess("", 0)
+    res.stdout = ""
+    if maintenance.VER_CMD == cmd:
+        res.stdout = version_res
+        return res
+    if fnmatch.fnmatch(cmd, maintenance.PRJ_CMD.format("*")):
+        if not any(prj in cmd for prj in valid_prj_list):
+            res.returncode = 1
+        return res
+    if fnmatch.fnmatch(cmd, maintenance.PER_MAINT_CMD.format("*")):
+        if "hpc-toolkit-test0" in cmd:
+            for vm in sched_maint_vms:
+                res.stdout += f"{vm}\n"
+        return res
+    if fnmatch.fnmatch(cmd, maintenance.UPC_MAINT_CMD.format("*")):
+        if "hpc-toolkit-test0" in cmd:
+            for vm in up_maint_vms:
+                res.stdout += f"{vm[0]}\t{vm[1]}\t{vm[2]}\t{vm[3]}\t{vm[4]}\n"
+        return res
+
+    return res
+
+@unittest.mock.patch('sys.stdout', new_callable=io.StringIO)
+@unittest.mock.patch('tools.maintenance.maintenance.subprocess.run',
+                     wraps=subprocess_replace)
+class TestProgram(unittest.TestCase):
+
+    per_test_str1 = "No nodes with periodic maintenance"
+    per_test_str2 = "Nodes with PERIODIC maintenance"
```

Comment:
nit: rename to `no_per_str` and `has_per_str` or something that makes it easer to read below which one does and does not have periodic maintenance.

Reply:
Done.